clients/java/signalr/src/test/java/HubConnectionTest.java

```diff
@@ -59,14 +57,134 @@ public void SendWithParamTriggersOnHandler() throws InterruptedException {
         assertEquals("Hello World", value.get());
     }
 
+    @Test
+    public void SendWithTwoParamsTriggersOnHandler() throws Exception {
+        AtomicReference<String> value1 = new AtomicReference<>();
+        AtomicReference<Double> value2 = new AtomicReference<>();
+
+        Transport mockTransport = new MockEchoTransport();
+        HubConnection hubConnection = new HubConnection("http://example.com", mockTransport);
+
+        hubConnection.on("inc", (param1, param2) ->{
+            assertNull(value1.get());
+            assertNull((value2.get()));
+
+            value1.set(param1);
+            value2.set(param2);
+        }, String.class, Double.class);
+
+        hubConnection.start();
+        hubConnection.send("inc", "Hello World", 12);
+
+        // Confirming that our handler was called and the correct message was passed in.
+        assertEquals("Hello World", value1.get());
+        assertEquals(12, value2.get(), 0);
+    }
+
+    @Test
+    public void SendWithThreeParamsTriggersOnHandler() throws Exception {
+        AtomicReference<String> value1 = new AtomicReference<>();
+        AtomicReference<String> value2 = new AtomicReference<>();
+        AtomicReference<String> value3 = new AtomicReference<>();
+
+        Transport mockTransport = new MockEchoTransport();
+        HubConnection hubConnection = new HubConnection("http://example.com", mockTransport);
+
+        hubConnection.on("inc", (param1, param2, param3) ->{
+            assertNull(value1.get());
+            assertNull(value2.get());
+            assertNull(value3.get());
+
+            value1.set(param1);
+            value2.set(param2);
+            value3.set(param3);
+        }, String.class, String.class, String.class);
+
+        hubConnection.start();
+        hubConnection.send("inc", "A", "B", "C");
+
+        // Confirming that our handler was called and the correct message was passed in.
+        assertEquals("A", value1.get());
+        assertEquals("B", value2.get());
+        assertEquals("C", value3.get());
+    }
+
+    @Test
+    public void SendWithFourParamsTriggersOnHandler() throws Exception {
+        AtomicReference<String> value1 = new AtomicReference<>();
```

Comment:
Is one of the reasons of using AtomicReference here to avoid an error saying that a variable _is not final or effectively final_?

Reply:
Yeah, that's the reason. is there a better way?

Reply:
I personally find this Java _feature_ limiting. What actually happens if an assert fails inside the lambda? Will the test fail correctly or will it crash the entire suite? I think it will not fail the test because you likely would have already left the method by the time you receive the callback. I wonder if you can actually assert the value inside the `on` method. 

Also, I find this test incorrect. I am not sure how this test does not fail because you never ensure that the callback was actually called before verifying the values. You also never stop the connection which you should. To properly stop the connection (since Java obviously does not have async test methods because it does not have async/await/Task) you need to kind of do what we did in the old SignalR. You use a `countdownLatch` which you release in the callback and which you await (`CountDownLatch.await`) with timeout before stopping the connection. You stop the connection in a `finally` because when the timeout expires it throws and you want to stop the connection regardless. If asserts inside the callback work as expected (i.e. a failed assert just fails the test) and you use the latch then you probably don't need to use AtomicReferences. You will not however be able to verify it the hub method was invoked more than once. You kind of do it now but this is really by accident because you don't close the connection and I don't really know what/how this would fail if you received the second call after the test has finished but the connection is still running on a background thread.

Or maybe AtomicReferences here are just fine (only I believe you still need the latch to block verification and closing the connection until you are really receive the method because what you have now either does not work at all or is super racy.